### ISSUE TITLE
Error when changing language

### ISSUE THREAD

**Issue by nmpetkov** [zikula/core#2320]
To reproduce:
1. Install 2 languages in the site.
2. Activate Language block.
3. Switch to non-default language.

On the address bar of the browser appere:
http://xyz.com/en
(in this case en is non-default language)

Here is image with the error:

![image](https://cloud.githubusercontent.com/assets/628801/6516297/665f39a4-c399-11e4-8ed8-c33c71fd21cd.png)

**Comment by craigh** [comment]
to get the full exception trace, please change the `custom_parameters.yml` values:

```yaml
    env: dev
    debug: true
```

**Comment by craigh** [comment]
ping @cmfcmf

**Comment by nmpetkov** [comment]
>  env: dev
>    debug: true
I can provide full trace.
The problem is related to routes, as I guess.

**Comment by Guite** [comment]
@nmpetkov: please do so (stack trace)
@cmfcmf: any idea which class should handle this / where to look at?

**Comment by nmpetkov** [comment]
```
[1] Symfony\Component\HttpKernel\Exception\NotFoundHttpException: Page not found.
    at n/a
        in /home/cmstory/public_html/zktest/lib/Zikula/Bundle/CoreBundle/EventListener/LegacyRouteListener.php line 80

    at Zikula\Bundle\CoreBundle\EventListener\LegacyRouteListener->onKernelRequest(object(GetResponseEvent), 'kernel.request', object(TraceableEventDispatcher))
        in  line 

    at call_user_func(array(object(LegacyRouteListener), 'onKernelRequest'), object(GetResponseEvent), 'kernel.request', object(TraceableEventDispatcher))
        in /home/cmstory/public_html/zktest/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/Debug/WrappedListener.php line 61

    at Symfony\Component\EventDispatcher\Debug\WrappedListener->__invoke(object(GetResponseEvent), 'kernel.request', object(Zikula_EventManager))
        in  line 

    at call_user_func(object(WrappedListener), object(GetResponseEvent), 'kernel.request', object(Zikula_EventManager))
        in /home/cmstory/public_html/zktest/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/EventDispatcher.php line 164

    at Symfony\Component\EventDispatcher\EventDispatcher->doDispatch(array(object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener)), 'kernel.request', object(GetResponseEvent))
        in /home/cmstory/public_html/zktest/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/EventDispatcher.php line 53

    at Symfony\Component\EventDispatcher\EventDispatcher->dispatch('kernel.request', object(GetResponseEvent))
        in /home/cmstory/public_html/zktest/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/ContainerAwareEventDispatcher.php line 167

    at Symfony\Component\EventDispatcher\ContainerAwareEventDispatcher->dispatch('kernel.request', object(GetResponseEvent), null)
        in /home/cmstory/public_html/zktest/lib/legacy/Zikula/EventManager.php line 99

    at Zikula_EventManager->dispatch('kernel.request', object(GetResponseEvent))
        in /home/cmstory/public_html/zktest/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/Debug/TraceableEventDispatcher.php line 112

    at Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher->dispatch('kernel.request', object(GetResponseEvent))
        in /home/cmstory/public_html/zktest/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php line 126

    at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Zikula_Request_Http), '1')
        in /home/cmstory/public_html/zktest/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php line 66

    at Symfony\Component\HttpKernel\HttpKernel->handle(object(Zikula_Request_Http), '1', true)
        in /home/cmstory/public_html/zktest/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/DependencyInjection/ContainerAwareHttpKernel.php line 64

    at Symfony\Component\HttpKernel\DependencyInjection\ContainerAwareHttpKernel->handle(object(Zikula_Request_Http), '1', true)
        in /home/cmstory/public_html/zktest/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/Kernel.php line 186

    at Symfony\Component\HttpKernel\Kernel->handle(object(Zikula_Request_Http))
        in /home/cmstory/public_html/zktest/index.php line 25

```

**Comment by Guite** [comment]
Proposal:

in https://github.com/zikula/core/blob/1.4/src/lib/Zikula/Bundle/CoreBundle/EventListener/LegacyRouteListener.php#L76

replace
```
    if ($path == "" || $path == "/") {
        // we have a static homepage
        $this->setResponse($event, new Response(''));
    } else {
        throw new NotFoundHttpException(__('Page not found.'));
    }
```

by
```
    $isAllowedNonModulePage = false;
    if ($path == '' || $path == '/') {
        $isAllowedNonModulePage = true;
    } else {
        $customentrypoint = System::getVar('entrypoint');
        $root = empty($customentrypoint) ? 'index.php' : $customentrypoint;

        // get base path to work out our current url
        $parsedURL = parse_url(System::getCurrentUri());
        // strip any unwanted content from the provided URL
        $tobestripped = array(System::getBaseUri(), "$root");
        $path = str_replace($tobestripped, '', $parsedURL['path']);
        $path = trim($path, '/');
        $args = explode('/', rtrim($path, '/'));

        $isAllowedNonModulePage = (ZLanguage::isLangParam($args[0]) && in_array($args[0], ZLanguage::getInstalledLanguages()));
    }

    if ($isAllowedNonModulePage) {
        // we have a static homepage
        $this->setResponse($event, new Response(''));
    } else {
        throw new NotFoundHttpException(__('Page not found.'));
    }
```

Please review.

**Comment by craigh** [comment]
I would wait for @cmfcmf on this. I think symfony routes are supposed to account for language more gracefully. Christian said he would look into some tickets in a couple days.

**Comment by craigh** [comment]
possibly refs #1985

**Comment by craigh** [comment]
shorturls enabled?

**Comment by nmpetkov** [comment]
Yes, shorturls  enabled.

**Comment by craigh** [comment]
what happens when not?

**Comment by nmpetkov** [comment]
Works as expected (```http://www.xyz.com/index.php?lang=en ```).

**Comment by cmfcmf** [comment]
I wonder why the RoutesListener tries to handle the `/en` url. Shouldn't the legacy listener run and detect it as valid shorturl?

**Comment by nmpetkov** [comment]
The problem seem to relate to #2348.

**Comment by craigh** [comment]
> I wonder why the RoutesListener tries to handle the `/en` url. Shouldn't the legacy listener run and detect it as valid shorturl?

I'm guessing because routes are filtered first. `mywebsite.com/en` looks like a route, so the router tries to match and fails.

**Comment by craigh** [comment]
I wonder if routes must be reloaded when the language is changed?

**Comment by craigh** [comment]
@nmpetkov did you click the `Reload multilingual routing settings` admin link in the Routes module? (see directions at top of routes list)

**Comment by craigh** [comment]
also - how do I get a translation to test with?

**Comment by nmpetkov** [comment]
I have just clicked on the link ```Reload multilingual routing settings``` in Routes module, and get ```Done! Routing settings renewed.``` to ensure procedure i successful, but the error remains when changing language.

**Comment by craigh** [comment]
what if you change the language in the settings and not with the block?

**Comment by nmpetkov** [comment]
In the settings you can only change default language, switching between languages appear in frontend of the site.

**Comment by craigh** [comment]
I understand that. What happens when you do that?

**Comment by nmpetkov** [comment]
Changing default language is without problems, no errors, as seen from pictures below:
![image](https://cloud.githubusercontent.com/assets/628801/6648174/4ec44db0-c9df-11e4-9f6f-c3d19121040e.png)

![image](https://cloud.githubusercontent.com/assets/628801/6648180/650103a2-c9df-11e4-93df-3779f2354b28.png)

**Comment by craigh** [comment]
and all the routes work properly?

**Comment by nmpetkov** [comment]
I'm preparing a translation to be able to test.

**Comment by nmpetkov** [comment]
Changing default language in settings work properly, but trying to install new language not. I'll try to see why and will post later.

**Comment by nmpetkov** [comment]
To install new language, for example ```de```, create file:
app/Resources/locale/de/locale.ini
with proper information, and files:
app/Resources/locale/de/LC_MESSAGES/zikula.po
app/Resources/locale/de/LC_MESSAGES/zikula.mo
app/Resources/locale/de/LC_MESSAGES/zikula_js.po
app/Resources/locale/de/LC_MESSAGES/zikula_js.mo

**Comment by craigh** [comment]
@cmfcmf I believe the problem may be somewhere in `\JMS\I18nRoutingBundle\Router\I18nLoader::load`

**Comment by craigh** [comment]
The original post here is mainly a problem with the HOMEPAGE only. While the core currently still has issues related to translated routes, most of these are corrected by #2353 
The homepage issue however is not corrected yet.

**Comment by craigh** [comment]
This is progress I think. It is definitely serving up this page in the selected language (German here) but it is (improperly) throwing a 404 for the homepage. So, mostly right, just don't want the error.

![screen shot 2015-03-16 at 5 29 16 pm](https://cloud.githubusercontent.com/assets/350048/6676936/04659d6c-cc02-11e4-991f-87e8e6ec3345.png)

**Comment by Guite** [comment]
Should be fixed by #2354. If you still experience problems please reopen.

**Comment by craigh** [comment]
this *may* also require #2353 - not sure.
